src/fromager/candidate.py

```diff
@@ -0,0 +1,81 @@
+import typing
+from email.message import EmailMessage, Message
+from email.parser import BytesParser
+from io import BytesIO
+from typing import TYPE_CHECKING
+from zipfile import ZipFile
+
+import requests
+from packaging.requirements import Requirement
+from packaging.utils import (
+    canonicalize_name,
+)
+from packaging.version import Version
+
+# fix for runtime errors caused by inheriting classes that are generic in stubs but not runtime
+# https://mypy.readthedocs.io/en/latest/runtime_troubles.html#using-classes-that-are-generic-in-stubs-but-not-at-runtime
+if TYPE_CHECKING:
+    Metadata = Message[str, str]
+else:
+    Metadata = Message
+
+
+class Candidate:
+    def __init__(
+        self,
+        name: str,
+        version: Version,
+        url: str,
+        extras: typing.Iterable[str] | None = None,
+        is_sdist: bool | None = None,
+    ):
+        self.name = canonicalize_name(name)
+        self.version = version
+        self.url = url
+        self.extras = extras
+        self.is_sdist = is_sdist
+
+        self._metadata: Metadata | None = None
+        self._dependencies: list[Requirement] | None = None
+
+    def __repr__(self) -> str:
+        if not self.extras:
+            return f"<{self.name}=={self.version}>"
+        return f"<{self.name}[{','.join(self.extras)}]=={self.version}>"
+
+    @property
+    def metadata(self) -> Metadata:
+        if self._metadata is None:
+            self._metadata = get_metadata_for_wheel(self.url)
+        return self._metadata
+
+    def _get_dependencies(self) -> typing.Iterable[Requirement]:
+        deps = self.metadata.get_all("Requires-Dist", [])
+        extras = self.extras if self.extras else [""]
+
+        for d in deps:
+            r = Requirement(d)
+            if r.marker is None:
+                yield r
+            else:
+                for e in extras:
+                    if r.marker.evaluate({"extra": e}):
+                        yield r
+
+    @property
+    def dependencies(self) -> list[Requirement]:
+        if self._dependencies is None:
+            self._dependencies = list(self._get_dependencies())
+        return self._dependencies
+
+
+def get_metadata_for_wheel(url: str) -> Metadata:
+    data = requests.get(url).content
```

Comment:
I just noticed that this is a problematic function. This line can download several GB of data and consume several GB of memory.

We should use https://peps.python.org/pep-0658/ if possible and use streaming IO if package metadata is not available.

Reply:
I will file a separate issue for this because the url being passed here is the download url and it doesn't have access to the tags that PEP 658 mentions. I will need to change some code in the resolver